table/tables/partition.go

```diff
@@ -714,9 +772,26 @@ func (lp *ForListPruning) locateListColumnsPartitionByRow(ctx sessionctx.Context
 	return location[0].PartIdx, nil
 }
 
-// buildListPartitionValueMap builds list columns partition value map for the specified column.
+// LocateRange locates partition range by the column range
+func (lp *ForListPruning) LocateRange(lowVal int64, isLowNull bool, highVal int64, isHighNull bool) []int {
```

Comment:
I am a little worried about correctness, so I prefer adding a unit test for this function.

I already add a unit test in my local:

1. change the methos interface to:
```go
// LocateRange locates partition range by the column range
func (lp *ForListPruning) LocateRange(ctx sessionctx.Context, r *ranger.Range) ([]int, error) {
	if r.LowVal[0].Kind() == types.KindMinNotNull {
		r.LowVal[0] = types.GetMinValue(lp.PruneExpr.GetType())
	}
	if r.HighVal[0].Kind() == types.KindMaxValue {
		r.HighVal[0] = types.GetMaxValue(lp.PruneExpr.GetType())
	}
	lowVal, isLowNull, err := lp.PruneExpr.EvalInt(ctx, chunk.MutRowFromDatums(r.LowVal).ToRow())
	if err != nil {
		return nil, err
	}
	highVal, isHighNull, err := lp.PruneExpr.EvalInt(ctx, chunk.MutRowFromDatums(r.HighVal).ToRow())
	if err != nil {
		return nil, err
	}
	if r.LowExclude {
		lowVal++
	}
	if !r.HighExclude {
		highVal++
	}
	partitionIdxes := make([]int, 0, lp.sorted.Len())
	if isLowNull {
		partitionIdxes = append(partitionIdxes, lp.nullPartitionIdx)
	}
	if isHighNull {
		partitionIdxes = append(partitionIdxes, lp.nullPartitionIdx)
	}
	lp.sorted.AscendRange(newBtreeSearchItem(lowVal), newBtreeSearchItem(highVal), func(item btree.Item) bool {
		partitionIdxes = append(partitionIdxes, item.(*btreeItem).partitionIdx)
		return true
	})
	return partitionIdxes, nil
}
```

2. put below unit test in `table/tables/partition_test.go`:

```go
// partitionTable is for those tables which implement partition.
type partitionTable interface {
	PartitionExpr() (*tables.PartitionExpr, error)
}

func (ts *testSuite) TestListPartitionRangePrune(c *C) {
	tk := testkit.NewTestKitWithInit(c, ts.store)
	tk.MustExec("use test")
	tk.MustExec("drop tables if exists t1")
	tk.MustExec("set @@session.tidb_enable_list_partition = ON")
	tk.MustExec(`create table t1 (id int, a int, b int) partition by list ( a ) (partition p0 values in (1,2,3,4,5), partition p1 values in (6,7,8,9,10), partition p2 values in (null));`)
	tb, err := ts.dom.InfoSchema().TableByName(model.NewCIStr("test"), model.NewCIStr("t1"))
	c.Assert(err, IsNil)
	partExpr, err := tb.(partitionTable).PartitionExpr()
	c.Assert(err, IsNil)
	cases := []struct {
		r      ranger.Range
		result []int
	}{
		{
			r:      ranger.Range{LowVal: types.MakeDatums(0), HighVal: types.MakeDatums(1)},
			result: []int{0},
		},
		{
			r:      ranger.Range{LowVal: types.MakeDatums(0), HighVal: types.MakeDatums(6)},
			result: []int{0, 0, 0, 0, 0, 1}, // need deduplicate the result, the expect should be {0,1}
		},
		{
			r:      ranger.Range{LowVal: []types.Datum{types.MinNotNullDatum()}, HighVal: []types.Datum{types.MaxValueDatum()}},
			result: []int{0, 0, 0, 0, 0, 1, 1, 1, 1, 1}, // need deduplicate the result, the expect should be {0,1}
		},
		{
			r:      ranger.Range{LowVal: types.MakeDatums(nil), HighVal: []types.Datum{types.MaxValueDatum()}},
			result: []int{2, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1}, // need deduplicate the result, the expect should be {0,1,2}
		},
		{
			r:      ranger.Range{LowVal: types.MakeDatums(nil), HighVal: []types.Datum{types.MaxValueDatum()}},
			result: []int{2, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1}, // need deduplicate the result, the expect should be {0,1,2}
		},
		{
			r:      ranger.Range{LowVal: types.MakeDatums(nil), HighVal: types.MakeDatums(nil)},
			result: []int{2, 2}, // need deduplicate the result, the expect should be {2}
		},
	}

	for _, ca := range cases {
		ret, err := partExpr.ForListPruning.LocateRange(tk.Se, &ca.r)
		c.Assert(err, IsNil)
		c.Assert(ret, DeepEquals, ca.result)
	}
}
```

Reply:
Done